src/main/java/net/neoforged/neoforge/transfer/handlers/wrappers/fluids/BlockFluidHandler.java

```diff
@@ -0,0 +1,197 @@
+
+/**
+ * A handler for placing and picking up fluid blocks in the world.
+ * <p>
+ * This is commonly used with the {@link FluidUtil#tryPickupFluidAsPlayer} or {@link FluidUtil#tryPlaceFluidAsPlayer}
+ */
+public class BlockFluidHandler extends SnapshotJournal<BlockFluidHandler.Snapshot> implements ISingleResourceHandler<FluidResource> {
+    private static final Logger LOGGER = LogManager.getLogger();
+
+    @Nullable
+    protected final Player player;
+    protected final Level level;
+    protected final BlockPos pos;
+    protected FluidStack fluidStack = FluidStack.EMPTY;
+
+    public BlockFluidHandler(@Nullable Player player, Level level, BlockPos pos) {
+        this.player = player;
+        this.level = level;
+        this.pos = pos;
+    }
+
+    public BlockFluidHandler(Level level, BlockPos pos) {
+        this(null, level, pos);
+    }
+
+    @Override
+    public FluidResource getResource(int index) {
+        Objects.checkIndex(index, size());
+        FluidState fluidState = level.getFluidState(pos);
+        return fluidState.getType().getDefaultResource();
+    }
+
+    @Override
+    public int getAmount(int index) {
+        Objects.checkIndex(index, size());
+        return getResource(index).isEmpty() ? 0 : FluidType.BUCKET_VOLUME;
+    }
+
+    @Override
+    public int getCapacity(int index, FluidResource resource) {
+        Objects.checkIndex(index, size());
+        //TODO Possibly check to see if fluid resource HAS a block state
+        return FluidType.BUCKET_VOLUME;
+    }
+
+    @Override
+    public boolean isValid(int index, FluidResource resource) {
+        Objects.checkIndex(index, size());
+        return true;
+    }
+
+    @Override
+    public boolean supportsInsertion() {
+        return true;
+    }
+
+    @Override
+    public boolean supportsExtraction() {
+        return true;
+    }
+
+    @Override
+    public int insert(FluidResource resource, int amount, TransactionContext transaction) {
+        if (ResourceHandlerUtil.isEmpty(resource, amount)) return 0;
+        if (amount < FluidType.BUCKET_VOLUME) return 0;
+        //We already are handling an insert this callback, we can't do more until closing.
+        // Something like vaporizing needs to be deferred until commit, but we don't have a
+        // system that allows us to queue callbacks as first-in-first-out.
+        if (!fluidStack.isEmpty()) return 0;
+
+        Fluid fluid = resource.getInstanceValue();
+        FluidStack fluidstack = resource.toStack(amount);
+        if (!fluid.getFluidType().canBePlacedInLevel(level, pos, fluidstack)) return 0;
+
+        BlockState state = level.getBlockState(pos);
+
+        boolean isFluidContainer = state.getBlock() instanceof LiquidBlockContainer container && container.canPlaceLiquid(null, level, pos, state, resource.getInstanceValue());
+        boolean isReplaceable = state.canBeReplaced(resource.getInstanceValue());
+
+        if (!isFluidContainer && !isReplaceable) return 0;
```

Comment:
I don't like these "hidden return" statements by tacking them on the end of the if statement line. Especially when there is a line of code on the next line